eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientInstantiationCheck.java

```diff
@@ -181,14 +201,206 @@ private void checkClassField(DetailAST objBlockToken) {
+        // If value of 'returns' is COLLECTION, and then log error if the return type of the method is not start with PagedFlux or Stream
+        if (SINGLE_RETURN_TYPE.equals(returnsAnnotationMemberValue)
+            && !returnType.startsWith(MONO) && !returnType.startsWith(RESPONSE)) {
+            log(methodDefToken, String.format(SINGLE_VALUE_RETURN_ERROR, SINGLE_RETURN_TYPE));
+        } else if (COLLECTION_RETURN_TYPE.equals(returnsAnnotationMemberValue)
+            && !returnType.startsWith(PAGED_FLUX) && !returnType.startsWith(STREAM)) {
+            log(methodDefToken, String.format(COLLECTION_RETURN_ERROR, COLLECTION_RETURN_TYPE));
+        }
+    }
+
+    /**
+     * Given the method is already annotated @ServiceMethod. Checks if the return type is Response or Mono<Response>,
+     * If the return type is Response, the method name must end with 'withResponse'.
+     * If the return type is Mono<Response>. the method name must not end with 'withResponse'.
+     *
+     * @param methodDefToken METHOD_DEF AST node
+     */
+    private void checkReturnTypeNamingPattern(DetailAST methodDefToken, String methodName) {
+        DetailAST typeToken = methodDefToken.findFirstToken(TokenTypes.TYPE);
+
+        StringBuilder sb = new StringBuilder();
+        getReturnType(typeToken, sb);
+        String returnType = sb.toString();
+
+        if (returnType.startsWith(RESPONSE_GENERIC_START) && !methodName.endsWith(WITH_RESPONSE)) {
+            log(methodDefToken, String.format("Return type is ''%s'', the method name must end with ''%s''.",
+                RESPONSE, WITH_RESPONSE));
+        } else if (returnType.equals(MONO_RESPONSE) && methodName.endsWith(WITH_RESPONSE)) {
+            log(methodDefToken, String.format("Return type is ''%s'', the method name must not end with ''%s''.",
+                MONO_RESPONSE, WITH_RESPONSE));
+        }
+    }
+
+    /**
+     * Checks the type Context should be in the right place.
+     * Synchronous method with annotation @ServiceMethod has to have {@code Context} as a parameter.
+     * Asynchronous method with annotation @ServiceMethod must not has {@code Context} as a parameter.
+     *
+     * @param methodDefToken METHOD_DEF AST token
+     */
+    private void checkContextInRightPlace(DetailAST methodDefToken) {
+        StringBuilder sb = new StringBuilder();
+        getReturnType(methodDefToken, sb);
+        DetailAST parametersToken = methodDefToken.findFirstToken(TokenTypes.PARAMETERS);
+        String returnType = sb.toString();
+
+        if (returnType.startsWith(RESPONSE_GENERIC_START) && !containsTypeParameter(parametersToken, CONTEXT)) {
+            log(methodDefToken, String.format("Synchronous method with annotation @ServiceMethod must has ''%s'' as a method parameter.", CONTEXT));
+        } else if (returnType.equals(MONO_RESPONSE) && containsTypeParameter(parametersToken, CONTEXT)) {
```

Comment:
This will not work. `MONO_RESPONSE` is defined as `Mono<Response>` but the return types are going to be `Mono<Response<T>>`. So, when you do `returnType.equals`, it won't match.

Also, you need a similar check for `PagedFlux` return type too.

Reply:
T here represents any type? And should I check PagedFlux<T> and PagedIterable<T> as well?